sql/parser/expr.go

```diff
@@ -241,20 +241,31 @@ func (NullVal) String() string {
 	return fmt.Sprintf("NULL")
 }
 
-// QualifiedName is a dot separated list of names.
-type QualifiedName []string
+// QualifiedName is a base name and an optional indirection expression.
+type QualifiedName struct {
+	Base     Name
+	Indirect Indirection
+}
 
 // Database returns the database portion of the name.
-func (n QualifiedName) Database() string {
-	if len(n) > 1 {
-		return n[0]
+func (n *QualifiedName) Database() string {
+	if len(n.Indirect) > 0 {
```

Comment:
No. While that would work with the current tests, I think it is only due to the way we always normalize the qualified name before using the `Database()` method. This logic here is that an expression like `foo` (with no indirection expression), `foo` represents a table or column name and not the database name. I'll add a comment to explain this.